### PR TITLE
Initial implementation of execution cache

### PR DESCRIPTION
**Pull request by mystenmark**
Initial implementation of the execution cache.

This passes test_full_node_cold_sync, and maybe some other simple tests.
It also allows the single node benchmark to run.

This is missing several features, most of which are noted with TODO(cache):
- We need to cache object locks (on both reads and writes)
- We need to flush data to disk before CheckpointExecutor finalizes a
  checkpoint, and remove dead objects from the cache at the same time
- Many traits that currently have impls only on AuthorityStore (e.g.
  BackingPackageStore, ChildObjectResolver, etc) must be implemented only on
  InMemoryCache instead.

### CODE REVIEW

File: crates/sui-storage/src/execution_cache.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    Deleted,
+}
+
+/// ExecutionCache is intended to provide an in-memory, write-behind or write-through cache used
+/// by transaction signing and execution.
+///
+/// Note that except where specified below, we do not provide any durability guarantees.
+/// Crash recovery is done by re-execution of transactions when necessary.
+#[async_trait]
+pub trait ExecutionCache {
```
Comment (halfprice) [inline crates/sui-storage/src/execution_cache.rs:25]: Is this an outdated version of the cache? It doesn't seem to be used anywhere. IIUC, the InMemoryCache with ExecutionCacheRead/ExecutionCacheWrite traits are the current execution cache implementation?

**Review by halfprice — COMMENTED** [review]
Hi Mark, saw that you are still working on this PR. I'll send out my two comments and wait when it's ready.

File: crates/sui-storage/src/execution_cache.rs (hunk 1 above)
Comment (mystenmark) [reply crates/sui-storage/src/execution_cache.rs:25]: yes, i need to delete this file, thanks for the reminder

File: crates/sui-core/src/in_mem_execution_cache.rs [hunk 2]
```diff
@@ -0,0 +1,840 @@
...
+            .notify(&tx_digest, &effects_digest);
+    }
+}
+
+// TODO(cache): this is not safe, because now we are committing persistent state to the db, while the rest
+// of the transaction outputs could simply be lost if the validator restarts. To fix this we must
+// cache locks (which we want to do anyway) but that's a big enough change for it to be worth
+// saving for a separate PR.
+#[instrument(level = "trace", skip_all)]
+fn write_locks(
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-core/src/in_mem_execution_cache.rs:1109]: At this point, new locks are always written synchronously and durably during execution. This is not correct during restarts as described in the comment.

My next PR adds correct lock caching.

File: crates/sui-core/src/authority.rs [hunk 3]
```diff
@@ -4594,6 +4649,48 @@ impl AuthorityState {
     }
 }
 
+impl AuthorityState {
+    /// Gets the input object keys from input object kinds, by determining the versions of owned,
+    /// shared and package objects.
+    /// When making changes, please see if check_sequenced_input_objects() below needs
```
Comment (halfprice) [inline crates/sui-core/src/authority.rs:4655]: This comment is no longer relevent?

File: crates/sui-core/src/in_mem_execution_cache.rs [hunk 4]
```diff
@@ -0,0 +1,1242 @@
...
+enum CacheResult<T> {
+    /// Entry is in the cache
+    Hit(T),
+    /// Entry is not in the cache and is known to not exist
+    NegativeHit,
+    /// Entry is not in the cache and may or may not exist in the store
+    Miss,
+}
+
+pub struct InMemoryCache {
```
Comment (lxfind) [inline crates/sui-core/src/in_mem_execution_cache.rs:452]: Probably better to make the naming more consistent: calling it `ExecutionCache` and etc. `InMemoryCache` doesn't really capture what's cached.

File: crates/sui-core/src/in_mem_execution_cache.rs [hunk 5]
```diff
@@ -0,0 +1,1242 @@
...
+    /// objects to the db in order to support past object queries on fullnodes.
+    /// When we move data into the object_cache we only retain the live objects.
+    ///
+    /// Further, we only remove objects in FIFO order, which ensures that the the cached
+    /// sequence of objects has no gaps. In other words, if we have versions 4, 8, 13 of
+    /// an object, we can deduce that version 9 does not exist. This also makes child object
+    /// reads efficient. `object_cache` cannot contain a more recent version of an object than
+    /// `objects`, and neither can have any gaps. Therefore if there is any object <= the version
+    /// bound for a child read in objects, it is the correct object to return.
+    objects: DashMap<ObjectID, BTreeMap<SequenceNumber, ObjectEntry>>,
```
Comment (lxfind) [inline crates/sui-core/src/in_mem_execution_cache.rs:466]: Probably need better naming for clarity.
One suggestion is to call "objects" as "all_dirty_objects", "objects_cache" as "cached_move_objects", and "packages" as "cached_move_packages".

File: crates/sui-core/src/authority/authority_store.rs [hunk 6]
```diff
@@ -1074,115 +825,57 @@ impl AuthorityStore {
...
 
-    /// Helper function for updating the objects and locks in the state
-    #[instrument(level = "trace", skip_all)]
-    async fn update_objects_and_locks(
+    /// Updates the state resulting from the execution of a certificate.
+    ///
+    /// Internally it checks that all locks for active inputs are at the correct
+    /// version, and then writes objects, certificates, parents and clean up locks atomically.
+    #[instrument(level = "debug", skip_all)]
+    pub async fn update_state(
```
Comment (halfprice) [inline crates/sui-core/src/authority/authority_store.rs:833]: This function also looks like not used? Will this function be the place where checkpoint executor updates the DB?

File: crates/sui-core/src/in_mem_execution_cache.rs (hunk 5 above)
Comment (lxfind) [reply crates/sui-core/src/in_mem_execution_cache.rs:466]: I also recommend partition the fields into two separate structs, one contain uncommitted results, and one contain cached data from DB.

File: crates/sui-core/src/in_mem_execution_cache.rs (hunk 4 above)
Comment (halfprice) [inline crates/sui-core/src/in_mem_execution_cache.rs:452]: I think one thing maybe worth documenting is that updating the fields in InMemoryCache doesn't require atomicity (IIUC based on our discussion).

**Review by lxfind — COMMENTED** [review]
I think we also need a TODO for the built-in indexing stuff.

**Review by halfprice — APPROVED** [review]
Left some minor comments. Look forward to the complete logic!

**Pull request closed without merging**